### PR TITLE
[WIP] Implement async dialect as compiler plugin

### PR DESCRIPTION
**Pull request by yijia1212**
* Implement async dialect including async.token, async.value types and async.await op
* Implement conversion from async dialect to async runtime module

### CODE REVIEW

File: async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+  MLIRContext *ctx = module->getContext();
+  SmallVector<Type> args{ValueType::get(ctx)};
+  SmallVector<Type> rets{IREE::Util::ObjectType::get(ctx)};
+  auto functionType = FunctionType::get(ctx, args, rets);
+
+  return addDecl(rewriter, module,
+                 StringAttr::get(ctx, "async.value.await.ref"), functionType);
+}
+
+class FuncOpConversion : public OpConversionPattern<func::FuncOp> {
```
Comment (ezhulenev) [inline async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp:107]: I believe that https://mlir.llvm.org/doxygen/namespacemlir.html#a12287a4adc9d89ce2ed49712b091c5fa does this already, there are many examples in MLIR passes

File: async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+      ConversionPatternRewriter &rewriter) const override {
+    if (!isa<ValueType>(op.getOperand().getType())) {
+      return rewriter.notifyMatchFailure(op, "unsupported awaitable type");
+    }
+    ModuleOp module = op->getParentOfType<ModuleOp>();
+    auto resultType = op.getResultType();
+    ImplicitLocOpBuilder b(op->getLoc(), rewriter);
+    if (resultType->isInteger(32)) {
+      auto funcOp = api->getValueAwaitI32(rewriter, module);
+      // auto callOp = b.create<func::CallOp>(funcOp.getSymName(), *resultType,
```
Comment (ezhulenev) [inline async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp:197]: leftover

File: async/compiler/src/openxla/compiler/Dialect/Async/IR/test/verify.mlir [hunk 3]
```diff
@@ -0,0 +1,21 @@
+// RUN: mlir-opt %s --iree-plugin=openxla-async -split-input-file -verify-diagnostics | FileCheck %s
```
Comment (ezhulenev) [inline async/compiler/src/openxla/compiler/Dialect/Async/IR/test/verify.mlir:1]: iree-opt, does it pass `ctest --test-dir build -R async`?

File: async/compiler/src/openxla/compiler/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+#include "openxla/compiler/Dialect/Async/IR/Async.h"
+#include "openxla/compiler/Dialect/Async/Transforms/Passes.h"
+
+#define GEN_PASS_DEF_ASYNCTOASYNCRUNTIME
+#include <iostream>
+
+#include "openxla/compiler/Dialect/Async/Transforms/Passes.h.inc"
+
+using namespace mlir;
+namespace IREE = mlir::iree_compiler::IREE;
```
Comment (ezhulenev) [inline async/compiler/src/openxla/compiler/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp:20]: `using namespace mlir::iree_compiler;` ? not sure why exactly I added this alias originally to cudnn project

File: async/compiler/src/openxla/compiler/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/Builders.h"
+#include "openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.h"
+#include "openxla/compiler/Dialect/Async/IR/Async.h"
+#include "openxla/compiler/Dialect/Async/Transforms/Passes.h"
+
+#define GEN_PASS_DEF_ASYNCTOASYNCRUNTIME
+#include <iostream>
```
Comment (ezhulenev) [inline async/compiler/src/openxla/compiler/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp:15]: move it to the top

File: async/compiler/src/openxla/compiler/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    : public ::impl::AsyncToAsyncRuntimeBase<AsyncToAsyncRuntimePass> {
+ public:
+  AsyncToAsyncRuntimePass() = default;
+  void runOnOperation() override;
+};
+
+}  // namespace
+
+//===----------------------------------------------------------------------===//
+void AsyncToAsyncRuntimePass::runOnOperation() {
```
Comment (ezhulenev) [inline async/compiler/src/openxla/compiler/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp:36]: newline

File: async/compiler/src/openxla/compiler/Dialect/Async/Transforms/test/async_to_runtime.mlir [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+// RUN:   --async-to-async-runtime | \
+// RUN: FileCheck %s
+
+func.func @await_token(%arg0: !async.token){
+  // CHECK: call @async.value.await.token
+  async.await %arg0 : !async.token
+  return
+}
+
+// CHECK:  func.func @await_token(%[[ARG:.*]]: !async.value) {
```
Comment (ezhulenev) [inline async/compiler/src/openxla/compiler/Dialect/Async/Transforms/test/async_to_runtime.mlir:11]: nit: indentation is slightly off, just one space between CHECK: and the first line

File: async/compiler/src/openxla/compiler/Dialect/Async/IR/Async.cpp [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+void ValueType::print(AsmPrinter &printer) const {
+  // Opaque async value type (`!async.value`)
+  if (!getValueType()) return;
+
+  printer << "<";
+  printer.printType(getValueType());
+  printer << '>';
+}
+
+Type ValueType::parse(mlir::AsmParser &parser) {
```
Comment (ezhulenev) [inline async/compiler/src/openxla/compiler/Dialect/Async/IR/Async.cpp:102]: Can't this be mostly auto generated? Couldn't find any example, can you please double check that we don't have anything better these days :)

**Review by ezhulenev — CHANGES_REQUESTED** [review]

File: async/compiler/src/openxla/compiler/Dialect/Async/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# Copyright 2023 The OpenXLA Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+iree_add_all_subdirs()
```
[the remark this answers is not in the document]
Comment (yijia1212) [reply async/compiler/src/openxla/compiler/Dialect/Async/CMakeLists.txt]: Fixed

File: async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+  func::FuncOp getTokenAwait(PatternRewriter &rewriter, ModuleOp module);
+  // Imports `@async.value.await.i32` into the module
+  func::FuncOp getValueAwaitI32(PatternRewriter &rewriter, ModuleOp module);
+  // Imports `@async.value.await.ref` into the module
+  func::FuncOp getValueAwaitRef(PatternRewriter &rewriter, ModuleOp module);
+
+  SymbolTable &symTable(ModuleOp module);
+
+  bool isScalarType(Type type) {
+    return type.isInteger(32) || type.isInteger(64) || type.isF32() ||
```
[the remark this answers is not in the document]
Comment (yijia1212) [reply async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp]: Fixed

File: async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp (hunk 1 above)
Comment (yijia1212) [reply async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp]: Fixed

File: async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+      : OpConversionPattern<T>(typeConverter, ctx), api(std::move(api)) {}
+
+  std::shared_ptr<AsyncAPI> api;
+};
+
+//===----------------------------------------------------------------------===//
+// Lowering for `async.await` with a token operand.
+//===----------------------------------------------------------------------===//
+
+struct ConvertTokenOp : public AsyncOpConversionPattern<AwaitOp> {
```
[the remark this answers is not in the document]
Comment (yijia1212) [reply async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp]: Fixed

File: async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp (hunk 2 above)
Comment (yijia1212) [reply async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp]: Fixed

File: async/compiler/src/openxla/compiler/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp (hunk 4 above)
Comment (yijia1212) [reply async/compiler/src/openxla/compiler/Dialect/Async/Transforms/AsyncToAsyncRuntime.cpp]: Fixed

File: async/compiler/src/openxla/compiler/Dialect/Async/Transforms/test/async_to_runtime.mlir (hunk 7 above)
Comment (yijia1212) [reply async/compiler/src/openxla/compiler/Dialect/Async/Transforms/test/async_to_runtime.mlir:11]: Fixed

File: async/iree_compiler_plugin_group.cmake [hunk 12]
```diff
@@ -0,0 +1,7 @@
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+add_subdirectory(${CMAKE_CURRENT_LIST_DIR}/compiler/src async/compiler)
```
Comment (ezhulenev) [inline async/iree_compiler_plugin_group.cmake:7]: missing newline at the end

Comment (yijia1212) [reply async/iree_compiler_plugin_group.cmake]: Fixed

File: async/compiler/src/openxla/compiler/Dialect/Async/IR/test/verify.mlir (hunk 3 above)
Comment (yijia1212) [reply async/compiler/src/openxla/compiler/Dialect/Async/IR/test/verify.mlir]: Passed

**Pull request closed without merging**